src/cargo/ops/cargo_compile.rs

```diff
@@ -1526,19 +1527,41 @@ fn find_named_targets<'a>(
     };
 
     if proposals.is_empty() {
-        let targets = packages.iter().flat_map(|pkg| {
-            pkg.targets()
-                .iter()
-                .filter(|target| is_expected_kind(target))
-        });
-        let suggestion = closest_msg(target_name, targets, |t| t.name());
-        anyhow::bail!(
-            "no {} target {} `{}`{}",
+        let targets = packages
+            .iter()
+            .flat_map(|pkg| {
+                pkg.targets()
+                    .iter()
+                    .filter(|target| is_expected_kind(target))
+            })
+            .collect::<Vec<_>>();
+        let suggestion = closest_msg(target_name, targets.iter(), |t| t.name());
+        if !suggestion.is_empty() {
+            anyhow::bail!(
+                "no {} target {} `{}`{}",
+                target_desc,
+                if is_glob { "matches pattern" } else { "named" },
+                target_name,
+                suggestion
+            );
+        }
+        let mut msg = String::new();
+        writeln!(
+            msg,
+            "no {} target {} `{}`.",
             target_desc,
             if is_glob { "matches pattern" } else { "named" },
             target_name,
-            suggestion
-        );
+        )?;
+        if targets.is_empty() {
+            writeln!(msg, "No {} targets available.", target_desc)?;
+            anyhow::bail!(msg);
+        }
+        writeln!(msg, "Available {} targets:", target_desc)?;
```

Comment:
nit: It looks like we are inconsistent on whether to capitalize this.  Let's lean towards not capitalizing for now

Reply:
It refers to:
```log
> cargo run --bin
error: "--bin" takes one argument.
Available binaries:
    abc
    xyz
```

So I prefer to keep it.